File: remote/src/main/java/com/kelsos/mbrc/interactors/library/GenreArtistInteractorImpl.java
package com.kelsos.mbrc.interactors.library;

import com.google.inject.Inject;
import com.kelsos.mbrc.domain.Artist;
import com.kelsos.mbrc.mappers.ArtistMapper;
import com.kelsos.mbrc.repository.library.ArtistRepository;
import java.util.List;
import rx.Observable;

public class GenreArtistInteractorImpl implements GenreArtistInteractor {
  @Inject private ArtistRepository repository;

  @Override public Observable<List<Artist>> getGenreArtists(long genreId) {
    return repository.getArtistsByGenreId(genreId)
        .flatMap(genreArtistViews -> Observable.just(ArtistMapper.mapGenreArtists(genreArtistViews)));
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/library/GenreArtistInteractorImpl.kt
package com.kelsos.mbrc.interactors.library

import com.kelsos.mbrc.domain.Artist
import com.kelsos.mbrc.mappers.ArtistMapper
import com.kelsos.mbrc.repository.library.ArtistRepository
import rx.Observable
import rx.lang.kotlin.toSingletonObservable
import javax.inject.Inject

class GenreArtistInteractorImpl
@Inject constructor(private val repository: ArtistRepository) : GenreArtistInteractor {

  override fun getGenreArtists(genreId: Long): Observable<List<Artist>> {
    return repository.getArtistsByGenreId(genreId)
        .flatMap {
          ArtistMapper.mapGenreArtists(it).toSingletonObservable()
        }
  }
}
